src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,459 @@
+        DateTimeOffset? timestamp = null,
+        int stackLevel = 1
+    )
+    {
+        timestamp ??= DateTimeOffset.UtcNow;
+        unit ??= MeasurementUnit.None;
+
+        Func<string, Metric> addValuesFactory = type switch
+        {
+            MetricType.Counter => _ => new CounterMetric(key, value, unit.Value, tags, timestamp),
+            MetricType.Gauge => _ => new GaugeMetric(key, value, unit.Value, tags, timestamp),
+            MetricType.Distribution => _ => new DistributionMetric(key, value, unit.Value, tags, timestamp),
+            MetricType.Set => _ => new SetMetric(key, (int)value, unit.Value, tags, timestamp),
+            _ => throw new ArgumentOutOfRangeException(nameof(type), type, "Unknown MetricType")
+        };
+
+        var timeBucket = Buckets.GetOrAdd(
+            timestamp.Value.GetTimeBucketKey(),
+            _ => new ConcurrentDictionary<string, Metric>()
+        );
+
+        timeBucket.AddOrUpdate(
+            GetMetricBucketKey(type, key, unit.Value, tags),
+            addValuesFactory,
+            (_, metric) =>
+            {
+                // This prevents multiple threads from trying to mutate the metric at the same time. The only other
+                // operations performed against metrics are adding one to the bucket (guaranteed to be atomic due to
+                // the use of a ConcurrentDictionary for the timeBucket) and removing buckets entirely.
+                //
+                // With a very small flushShift (e.g. 0.0) it might be possible for a metric to be emitted to a bucket
+                // that was removed after a flush, in which case that metric.Add(value) would never make it to Sentry.
+                // We've never seen this happen in unit testing (where we always set the flushShift to 0.0) so this
+                // remains only a theoretical possibility of data loss (not confirmed). If this becomes a real problem
+                // and we need to guarantee delivery of every metric.Add, we'll need to build a more complex mechanism
+                // to coordinate flushing with emission.
+                lock(metric)
+                {
+                    metric.Add(value);
+                }
+                return metric;
+            });
+
+        if (_options.ExperimentalMetrics is { EnableCodeLocations: true })
+        {
+            RecordCodeLocation(type, key, unit.Value, stackLevel + 1, timestamp.Value);
+        }
+    }
+
+    private readonly SemaphoreSlim _codeLocationLock = new(1,1);
+
+    internal void RecordCodeLocation(
+        MetricType type,
+        string key,
+        MeasurementUnit unit,
+        int stackLevel,
+        DateTimeOffset timestamp
+    )
+    {
+        var startOfDay = timestamp.GetDayBucketKey();
+        var metaKey = new MetricResourceIdentifier(type, key, unit);
+
+        var seenToday = _seenLocations.GetOrAdd(startOfDay,_ => []);
+        if (seenToday.Contains(metaKey))
```

Comment:
You're accessing the map outside of the lock and this will throw if one thread is on line 195 and another is here.

Reply:
I've expanded the scope of the lock... should be OK now I think.